sources/light_client.move

```diff
@@ -104,57 +112,109 @@ public fun new_light_client(params: Params, start_height: u64, start_headers: ve
+    let current_block = c.get_light_block_by_hash(current_block_hash);
     let current_header = current_block.header();
 
     // verify new header
     assert!(current_header.block_hash() == next_header.prev_block(), EBlockHashNotMatch);
     let next_block_difficulty = calc_next_required_difficulty(c, current_block, 0);
     assert!(next_block_difficulty == next_header.bits(), EDifficultyNotMatch);
 
+
     // https://learnmeabitcoin.com/technical/block/time
     // we only check the case "A timestamp greater than the median time of the last 11 blocks".
     // because  network adjusted time requires a miners local time.
-    let median_time = calc_past_median_time(c, current_block);
-    assert!(next_header.timestamp() > median_time, ETimeTooOld);
+    let median_time = c.calc_past_median_time(current_block);
+    assert!(next_header.timestamp() >= median_time, ETimeTooOld);
     next_header.pow_check();
 
     // update new header
     let next_height = current_block.height() + 1;
-    let next_light_block = new_light_block(next_height, raw_header, ctx);
+    let next_chain_work = current_block.chain_work() + next_header.calc_work();
+    let next_light_block = new_light_block(next_height, next_header, next_chain_work);
+
     c.finalized_height = next_height;
-    c.set_light_block(next_light_block);
+
+    c.add_light_block(next_light_block);
+    c.set_block_header_by_height(next_height, next_header);
+    next_header.block_hash()
+}
+
+fun extend_chain_from_block(c: &mut LightClient, start_point: vector<u8>, raw_headers: vector<vector<u8>>): vector<u8> {
+    let mut previous_block_hash = start_point;
+    raw_headers.do!(|raw_header| {
+        let header = new_block_header(raw_header);
+        previous_block_hash = c.insert_header(previous_block_hash, header);
+    });
+
+    previous_block_hash
 }
 
+public entry fun insert_headers(c: &mut LightClient, raw_headers: vector<vector<u8>>) {
+    // TODO: check if we can use BlockHeader instead of raw_header or vector<u8>(bytes)
+    assert!(!raw_headers.is_empty(), EHeaderListIsEmpty);
+
+    let first_header = new_block_header(raw_headers[0]);
+    let latest_block_hash = c.latest_block().header().block_hash();
+
+    if (first_header.prev_block() == latest_block_hash) {
+        // extend current fork
+       c.extend_chain_from_block(first_header.prev_block(), raw_headers);
+    } else {
+        // handle a fork choice
+
+        assert!(c.exist(first_header.prev_block()), EBlockDoesnotExist);
+        let current_chain_work = c.latest_block().chain_work();
+        let candidate_fork_head_hash = c.extend_chain_from_block(first_header.prev_block(), raw_headers);
+
+        let candidate_head = c.get_light_block_by_hash(candidate_fork_head_hash);
+        let candidate_chain_work= candidate_head.chain_work();
+        assert!(current_chain_work < candidate_chain_work, EForkNotEnoughPower);
+        c.rollback(first_header.block_hash(), candidate_head.header().block_hash());
```

Comment:
This is a bug!  
The right rollback should be from the old forkhead to the parent block of the first_header.